expression/builtin_arithmetic.go

```diff
@@ -601,35 +597,31 @@ func (s *builtinArithmeticIntDivideIntUnsignedSig) evalInt(row []types.Datum) (v
 		return 0, isNull, errors.Trace(err)
 	}
 
-	if a < 0 && -a >= b {
-		return 0, false, types.ErrOverflow.GenByArgs("BIGINT", fmt.Sprintf("(%d, %d)", a, b))
-	}
-
-	if a != 0 && b < 0 && -b <= a {
-		return 0, false, types.ErrOverflow.GenByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%d, %d)", a, b))
-	}
-
-	return a / b, false, nil
-}
-
-func (s *builtinArithmeticIntDivideIntSig) evalInt(row []types.Datum) (val int64, isNull bool, err error) {
-	sc := s.ctx.GetSessionVars().StmtCtx
-	b, isNull, err := s.args[1].EvalInt(row, sc)
-	if isNull || err != nil {
-		return 0, isNull, errors.Trace(err)
-	}
+	var (
+		ret int64
+		val uint64
+	)
+	isLHSUnsigned := mysql.HasUnsignedFlag(s.args[0].GetType().Flag)
+	isRHSUnsigned := mysql.HasUnsignedFlag(s.args[1].GetType().Flag)
 
-	if b == 0 {
-		return 0, true, nil
+	switch {
+	case isLHSUnsigned && isRHSUnsigned:
+		ret = a / b
+	case isLHSUnsigned && !isRHSUnsigned:
+		val, err = types.DivUintWithInt(uint64(a), b)
+		ret = int64(val)
+	case !isLHSUnsigned && isRHSUnsigned:
+		val, err = types.DivIntWithUint(a, uint64(b))
+		ret = int64(val)
+	case !isLHSUnsigned && !isRHSUnsigned:
+		ret, err = types.DivInt64(a, b)
 	}
 
-	a, isNull, err := s.args[0].EvalInt(row, sc)
-	if isNull || err != nil {
-		return 0, isNull, errors.Trace(err)
+	if err != nil {
```

Comment:
useless check,
use `return ret, false, errors.Trace(err)` is enough,
if err == nil, errors.Trace(err) returns nil.